semantics/small-steps-test/test/Control/State/Transition/Examples/Modular/Base.hs

```diff
@@ -0,0 +1,359 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE ViewPatterns #-}
+
+-- | Base for a modular STS example
+module Control.State.Transition.Examples.Modular.Base where
+
+import Control.State.Transition
+import qualified Control.State.Transition.Examples.Modular.Base.Types as BT
+import Data.Group (Group, invert)
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import Data.Maybe (isJust)
+import Data.Monoid (Sum (..))
+import Optics
+
+-------------------------------------------------------------------------------
+-- Lens stuff
+-------------------------------------------------------------------------------
+class (HasGetter a b, HasSetter a b) => HasLens a b where
+  lensy :: Lens' a b
+  lensy = lens (view gettery) (flip $ set settery)
+
+extract :: HasGetter a b => a -> b
+extract = view gettery
+
+inject :: HasSetter a b => b -> a -> a
+inject = set settery
+
+class HasGetter a b where
+  gettery :: Getter a b
+
+class HasSetter a b where
+  settery :: Setter' a b
+
+instance HasGetter a a where
+  gettery = castOptic equality
+
+instance HasSetter a a where
+  settery = castOptic equality
+
+instance HasLens a a where
+  lensy = castOptic equality
+
+-------------------------------------------------------------------------------
+-- Era_0
+-------------------------------------------------------------------------------
+
+data Era_0
+
+newtype Coin = Coin Int
+  deriving newtype (Eq, Ord, Show)
+  deriving
+    (Group, Semigroup, Monoid)
+    via Sum Int
+
+type instance BT.Coin Era_0 = Coin
+
+newtype AccountName = AccountName String
+  deriving newtype (Eq, Ord, Show)
+
+type instance BT.AccountName Era_0 = AccountName
+
+data Account e = Account
+  { name :: BT.AccountName e,
+    balance :: BT.Coin e
+  }
+
+deriving stock instance
+  ( Eq (BT.AccountName e),
+    Eq (BT.Coin e)
+  ) =>
+  Eq (Account e)
+
+deriving stock instance
+  ( Show (BT.AccountName e),
+    Show (BT.Coin e)
+  ) =>
+  Show (Account e)
+
+type instance BT.Account Era_0 = Account Era_0
+
+data Tx e = Tx
+  { fromAcnt :: BT.AccountName e,
+    toAcnt :: BT.AccountName e,
+    amount :: BT.Coin e
+  }
+
+type instance BT.Tx Era_0 = Tx Era_0
+
+data Admin e
+  = OpenAccount (BT.AccountName e) (BT.Coin e)
+  | Withdraw (BT.AccountName e) (BT.Coin e)
+  | CloseAccount (BT.AccountName e)
+
+type instance BT.Admin Era_0 = Admin Era_0
+
+data Op e
+  = OpTx (BT.Tx e)
+  | OpAdmin (BT.Admin e)
+
+type instance BT.Op Era_0 = Op Era_0
+
+newtype Accounting e = Accounting
+  { unAccounting :: Map (BT.AccountName e) (BT.Account e)
```

Comment:
Do the elements of the map also contain an `AccountName`? If so, is an invariant of this type that the account name of a key should be the same as the account name of the element at that key? Also if this is correct, why does `Account` need the account name as well?

Reply:
They do, it is, and because this ledger was only an example and I didn't stop to make it nice :-D